src/values/instruction_value.rs

```diff
@@ -684,16 +688,42 @@ impl<'ctx> InstructionValue<'ctx> {
         }
 
         let is_basic_block = unsafe { !LLVMIsABasicBlock(operand).is_null() };
-
         if is_basic_block {
             let bb = unsafe { BasicBlock::new(LLVMValueAsBasicBlock(operand)) };
 
-            Some(Operand::Block(bb.expect("BasicBlock should always be valid")))
-        } else {
-            Some(Operand::Value(unsafe { BasicValueEnum::new(operand) }))
+            return Some(Operand::Block(bb.expect("BasicBlock should always be valid")));
+        }
+
+        match self.try_ingest_metadata(operand) {
+            // This is indeed metadata.
+            Ok(Some(metadata)) => return Some(Operand::Metadata(metadata)),
+            // This is null metadata, which we can't convert to an operand.
+            Err(_) => return None,
+            // Not metadata at all.
+            Ok(None) => {}
         }
+
+        Some(Operand::Value(unsafe { BasicValueEnum::new(operand) }))
     }
 
+    #[llvm_versions(17..)] // LLVMIsAValueAsMetadata was introduced un 17.0.
+    unsafe fn try_ingest_metadata(self, operand: LLVMValueRef) -> Result<Option<MetadataValue<'ctx>>, ()> {
```

Comment:
There's really no reason to have this function. You can just cfg the match statement in get_operand_unchecked. This adds unnecessary complexity.